Client SDK Digital Marketplace Application Use Case (#2106)
Nexmo/nexmo-developer: _use_cases/digital-marketplace-client-sdk.md

## Make it your own

If you’d like to build your own version of this project to experiment with, you can [remix the Glitch](https://glitch.com/edit/#!/remix/green-crowberry) or [clone the GitHub repository](https://github.com/nexmo-community/client-sdk-marketplace-use-case) 

You also need to do the following:

- [Sign up for a Nexmo Account](https://dashboard.nexmo.com/sign-up) (take note of your API Key and API secret)

![Nexmo Developer Dashboard screenshot](/assets/screenshots/use-cases/digital-marketplace-client-sdk/api-key-and-secret.png)

- [Create an Application](https://dashboard.nexmo.com/voice/your-applications). An [Application](/application/overview) is a container for your project’s security and configuration details.

![Nexmo Developer Create An Application screenshot](/assets/screenshots/use-cases/digital-marketplace-client-sdk/create-an-application.png)

Note: We will not be using Event URL and Answer URL, so you can enter `https://example.com/event` and `https://example.com/answer` respectively.

Click “Generate public/private key pair” to generate your public key and download the `private.key` file to your computer.

Take note of your Application ID. 

> ### Using GitHub

Reviewer: I would add a L2 heading here called "Authenticating your application" and put the "Using GitHub" and "Using Glitch" L3 headings underneath it. It helps break up the page and make the side nav more useful.

Author: Yeah, totally agree. I noticed that the side navigation was determined by the headings after I did the pull request.

Move the `private.key` file to the root of your project:

![Nexmo Application private key location local screenshot](/assets/screenshots/use-cases/digital-marketplace-client-sdk/private-key-location-local.png)

### Using Glitch

Open the `private.key` file in a text editor. Then, in your Glitch project, create the file `/.data/private.key` and copy and paste in the contents of the `private.key`.

![Nexmo Application private key location Glitch screenshot](/assets/screenshots/use-cases/digital-marketplace-client-sdk/private-key-location-glitch.png)

> Regardless of whether you are remixing the Glitch or cloning the GitHub repository, you must configure the application using the `.env` file. Fill in each setting using the values you noted in the preceding steps.

Reviewer: Add a L2 heading here called "Configuring your application" and put the L3 "Using GitHub" and "Using Glitch" headings beneath it.

Author: Done

> (Note, if your .env file is empty, it should like this)

Reviewer: ```suggestion
The structure of the `.env` file is slightly different depending on whether you are using GitHub or Glitch:
```

Author: Going to add this in my next push. Not sure if I commit it here if it will cause a merge conflict when I try to push my other changes. Just to be safe.

### For GitHub

```
API_KEY="your-value-here"
API_SECRET="your-value-here"
APP_ID="your-value-here"
PRIVATE_KEY="/private.key"
```

### For Glitch

```
DANGEROUSLY_DISABLE_HOST_CHECK=true
API_KEY="your-value-here"
API_SECRET="your-value-here"
APP_ID="your-value-here"
PRIVATE_KEY="/.data/private.key"
```

Now we should be ready to go!

## Walkthrough

### Login

![Marketplace App login screenshot](/assets/screenshots/use-cases/digital-marketplace-client-sdk/app-login.png)

> The user will enter a username and select either a Seller or Buyer role.

Reviewer: ```suggestion
The user enters a username and select either the Seller or Buyer role.
```

Author: Will be added in next push.

> (Note: The request body has properties we can use for setting the user name, display name and image URL, but there is no such property for specifying the role. Fortunately, we can add our own properties in `custom_data`, so we'll create `role` in there.)

Reviewer: ```suggestion
The request body has properties we can use for setting the user name, display name and image URL, but there is no such property for specifying the role. Fortunately, we can add our own properties in `custom_data`, so we'll create `role` in there:
```

Author: Done.

*NexmoMarketplaceApp.js*

```jsx
  const submitUser = async (e) => {
    try{
      const results = await fetch('/createUser', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          name: name.split(' ').join('-'),
          display_name: name.trim(),
          image_url: `https://robohash.org/${name.split(' ').join('-')}`,
          properties: {
            custom_data: {
              "role": role
            }
          }
        })
      });
      await results.json();

>       // const data = await results.json();

Reviewer: Should we remove these commented-out lines for the tutorial?

Author: Great catch! An artifact of some refactoring.

      // setUserId(data.id);
      await login();
    } catch(err){
      console.log('getJWT error: ',err);
    }
  };

  // Get JWT to authenticate user
  const getJWT = async () => {
    try{
      const results = await fetch('/getJWT', {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({
          name: name.split(' ').join('-')
        })
      });
      const data = await results.json();
      return data.jwt;
    } catch(err){
      console.log('getJWT error: ',err);
    }
  };

  // Log in the user
  const login = async () => {
    try{
      const userJWT = await getJWT();
      const app =  await new NexmoClient({ debug: false }).login(userJWT);
      setNexmoApp(app);
      await getConversations();
      setStage('listings');
    } catch(err){
      console.log('login error: ',err);
    }
  };
```

> 

Reviewer: Add a L3 "Authentication" heading here

Author: Got it.

> Authentication is handled with [JWT](/concepts/guides/authentication#json-web-tokens-jwt). The application makes a call to the Node Express server to retrieve the JWT and then logs the user in.

Reviewer: ```suggestion
The Client SDK authenticates using [JWTs](/concepts/guides/authentication#json-web-tokens-jwt). The application makes a call to the Node Express server to retrieve the JWT and then logs the user in.
```

Author: Done.

    const jwt = nexmo.generateJwt({
        application_id: process.env.APP_ID,
        sub: req.body.name,
        exp: Math.round(new Date().getTime()/1000)+86400,
        acl: {
            "paths": {
                "/*/users/**":{},
                "/*/conversations/**":{},
                "/*/sessions/**":{},
                "/*/devices/**":{},
                "/*/image/**":{},
                "/*/media/**":{},
                "/*/applications/**":{},
                "/*/push/**":{},
                "/*/knocking/**":{}
            }
        }
    });
    res.send({jwt: jwt});
});

// the client calls this endpoint to create a new user in the Nexmo application,
// passing it a username and optional display name
app.post('/createUser', function(req, res) {
    console.log('/createUser: ',req);
    nexmo.users.create({
        name: req.body.name,
        display_name: req.body.display_name || req.body.name,
        image_url: req.body.image_url,
        properties: req.body.properties
    },(err, response) => {
        if (err) {
            res.sendStatus(500);
        } else {
            res.send({id: response.id});
        }
    });
});
```

> When the user is logged in, we retrieve a list of all the items for sale.

Reviewer: Add L3 heading here "Displaying items for sale"

Author: Got it.

*NexmoMarketplaceApp.js*

```jsx
  // Get all conversations, even the ones the user isn't a member of, yet.
  const getConversations = async() => {
    try{
        const results = await fetch('/getConversations', {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json'
          },
          body: JSON.stringify({
            page_size: 100
          })
        });
        const data = await results.json();
        setItems(data.conversations);
    } catch(err) {
      console.log('getConversations error: ',err);
    }
  };
```

*server.js*

```js
app.post('/getConversations', function(req, res) {
    console.log('/getConversations: ',req);
    nexmo.conversations.get({page_size: req.body.page_size},(err, response) => {
        if (err) {
            res.sendStatus(500);
        } else {
            res.send(response._embedded);
        }
    });
});
```

> ### Listings

Reviewer: ```suggestion
### Listing a new item for sale
```

Author: Noted.

We can alert the application that a new item has been listed for sale using a custom event called `item_details`, which passes the item details to the handler.

*NexmoMarketplaceApp.js*

```jsx
  const createConversation = async() => {
    try{
      const conversation = await nexmoApp.newConversation({
        name: itemName.split(' ').join('-'), // comment out to get a GUID
        display_name: itemName.trim(),
        properties:{
          custom_data:{
            title: itemName,
            description: itemDescription,
            price: itemPrice,
            image_url: itemImage,
          }
        }
      });
      await conversation.join();
      await conversation.sendCustomEvent({ type: 'custom:item_details', body: { title: itemName, description: itemDescription, price: itemPrice, image_url: itemImage }})
      await getConversations();
      setItemName('');
      setItemImage('');
      setItemDescription('');
      setItemPrice('');
    } catch(err){
      console.log('createConversation error: ',err);
    }
  };
```

With that, we then get an updated list with your item at the top.

Go ahead and click on your item.

> ### Item details page

Reviewer: ```suggestion
### The item details page
```

Author: Got it.

> Clicking on an item calls the Client SDK’s `getConversation` function.  The code checks to see if the current user is a Member of the Conversation. If not, the User is  joined to the Conversation.

Reviewer: ```suggestion
Clicking on an item calls the Client SDK’s `getConversation` function.  The code checks to see if the current user is a Member of the Conversation. If not, it adds the User as a Member.
```

Author: Done.

      for(const [,event] of allEvents.items) {
        let user = await nexmoApp.getUser(conversation.members.get(event.from).user.id);
        switch(event.type){
          case 'text':
            setChatMessages(chatMessages => [...chatMessages,{avatar: user.image_url, sender:conversation.members.get(event.from), message:event, me:conversation.me}]);
            break;
          case 'custom:item_details':
            setConversationItem({...conversationItem,...event.body, seller: user});
            break;
          case 'custom:stripe_payment':
            setChatMessages(chatMessages => [...chatMessages,{avatar: '', sender:{user:{name:'Stripe'}}, message:{body:{text:`${event.body.paymentDetails.description}: ${event.body.paymentDetails.status}`}}, me:''}]);
            if (event.body.paymentDetails.status === 'succeeded'){
              setConversationItem(prevState => {
                return { ...prevState, status: 'Sold' }
              });
            }
            break;
          default:
        }
      }
      setStage('conversation');
    } catch(err){
      console.log('getConversation error: ',err);
    }
  };
```

> Let’s say you want to purchase the item. When you click the Pay Now button, we raise another custom event (stripe_payment) with the Nexmo Client SDK. Note: In this use case, we simply mock the response from Stripe and leave the implementation of a payment gateway to you.

Reviewer: Add new L3 heading here "Purchasing items"

Author: Got it.

      if (event.body.paymentDetails.status === 'succeeded'){
        setConversationItem(prevState => {
          return { ...prevState, status: 'Sold' }
        });
      }
    };
    if(nexmoConversation){
      nexmoConversation.on('custom:stripe_payment', setStripePayment);
      return () => {
        nexmoConversation.off('custom:stripe_payment', setStripePayment);
      };
    }
  });
```

The listener displays the payment notification as a chat message. If the payment was “successful” we update the status of the Item to Sold and update the UI.

## Conclusion

With this Use Case, we demonstrate how to use the Nexmo Client SDK to send custom events and then listen for those events to update the state of the application.

## Where Next?

You will definitely want to add proper authentication if you are using this example as the basis for your production application. 

You might also want to consider adding more custom events to make the buying and selling experience a better one for your users. Perhaps you could allow users to add items they are interested in purchasing to a list of favorites? Or enable sellers to edit an item that they have listed for sale?

> _Useful Client SDK links:_

Reviewer 2: Make this a `##` - not italics.

Author: Got it.

> Overview: [https://developer.nexmo.com/client-sdk/overview](https://developer.nexmo.com/client-sdk/overview)

Author: Done.
